### PR TITLE
🚧 Portfolio Upgrade v1.1 (WIP)

### PR DESCRIPTION
**Pull request by thomasogara**
Upgrading the Portfolio to v1.1 (WIP)

### Change List:

* The home page now has a dynamic biographical section, as opposed to the previous version's static biography
* There is now a new page, gallery.html, which will host pages relating to the projects I create during the STEP internship. I am in the process of setting up a system for easily updating the project entries so that I don't have to edit the HTML in order to add more projects to the page, or to update existing projects.

### CODE REVIEW

File: portfolio/src/main/webapp/style.css [hunk 1]
```diff
@@ -8,10 +8,21 @@
   width: 650px;
 }
 
-#greeting-container {
-  margin-top: 20px;
+#biography-output{
```
[the remark this answers is not in the document]
Comment (thomasogara) [reply portfolio/src/main/webapp/style.css]: cool, sorted that out 👍

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -13,21 +13,89 @@
...
-        'I once ran arch linux on Windows using WSL for the lolz!',
-        'My dogs\' names are Polly and Rex!',
-        'My \'to-read\' list is currently growing at a rate of about 2bpm (books per month)!',
-      ];
+const slowFillBiography = async () => {
+  const biography_text = 
+      'Hi! My name\'s Thomas, and I\'m just about to start my third year of ' +
+      'studies in University College Dublin. I\'m currently a STEP intern ' +
+      'at Google, and this portfolio aims to showcase my contributions ' + 
+      'during my time here.';
```
Comment (magurmach) [inline portfolio/src/main/webapp/script.js:24]: I believe you can create multiline string just by using backtick (`). Just FYI.

File: portfolio/src/main/webapp/gallery-script.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Load project details from local registry of projects (projects.json)
+ */
+const loadProjects = async() => {
+  const request = new XMLHttpRequest();
+  request.open("GET", "projects.json", true);
```
Comment (magurmach) [inline portfolio/src/main/webapp/gallery-script.js:6]: Why do you need to make HTTP request to access local file? Can't you directly access it?

File: portfolio/src/main/webapp/gallery-script.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    json will always contain a top-level object which has a single member.
+    this member will be an array of objects. these objects will have three properties:
+    - title : a string containing the title of the project
+    - URL : a string which contains a URL linking to the blog post about
+        the project
+    - preview : a string which contains a URL linking to the preview image
+        for this project
+  */
+  const data = JSON.parse(json);
+  data.projects.map(async (project) => {
```
Comment (magurmach) [inline portfolio/src/main/webapp/gallery-script.js:27]: I am not clear on what are you trying to achieve here. You are trying to Map the json data to what? It's not easily readable.

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -13,21 +13,89 @@
...
+  let cursor_attached = false;
 
-  // Add it to the page.
-  const factContainer = document.getElementById('fact-container');
-  factContainer.innerText = fact;
+  /* 
+    wait_function must be non-null.
+    all null inputs are normalised to a known default.
+   */
+  if (wait_function == null) wait_function = function() {0};
```
Comment (magurmach) [inline portfolio/src/main/webapp/script.js:50]: 0 doesn't look like a good default for wait_function. I will say, as your original invocation uses (Math.floor(Math.random() * 50) + 50) which means the minimum is 50, default to 50.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -13,21 +13,89 @@
...
+                  substr(0, entity.innerText.length - 1);
+      cursor_attached = !cursor_attached;
+    }
+  };
+  const attachCursor = () => {
+    entity.innerText += cursor_attached ? '' : '|';
+    cursor_attached = !cursor_attached;
+  };
+  
+  for(let word of text.split(' ')){
```
Comment (magurmach) [inline portfolio/src/main/webapp/script.js:66]: I am not clear what is the goal here. Comment with the necessary details? Or attach a screenshot of the goal?

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js:24]: Yeah I saw that in the style doc, but the instructions also said that multi-line string literals could be implemented using just concatenation (https://google.github.io/styleguide/jsguide.html#features-strings-no-line-continuations). There was no clear indication as to which was preferable, so I wasn't sure which to use. I'll change this now though 👍

File: portfolio/src/main/webapp/gallery-script.js (hunk 3 above)
Comment (thomasogara) [reply portfolio/src/main/webapp/gallery-script.js:6]: We're only using client-side code for the portfoli,o at least at the moment, so this seemed like the easiest way to get the file onto the client's machine. I didn't know of any other way to include the file without making a second request once the javascript loads.

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js]: I've fixed this now, I think the suggestion of a 50ms wait makes much more sense 👍

File: portfolio/src/main/webapp/gallery-script.js (hunk 4 above)
Comment (thomasogara) [reply portfolio/src/main/webapp/gallery-script.js]: I've updated the docs now 👍 https://github.com/thomasogara/step-repository/pull/2/commits/5a27a88da131033ecf7d5279c4715f4b574eebdd

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js:84]: I've improved the documentation now 👍 Let me know if it needs more work https://github.com/thomasogara/step-repository/pull/2/commits/1411993e99833c3eac58a7a59506bd1cef8ed5d0

Comment (thomasogara) [reply portfolio/src/main/webapp/script.js:89]: I also added documentation for the nested for loops, just to better explain the intent of the function 👍 https://github.com/thomasogara/step-repository/pull/2/commits/d9e27cc5b6d81a0b38efc9feb1f7be6f610df1db

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -13,21 +13,118 @@
...
+ * @param text The text to write to the text field of the entity.
+ * @param wait_function A function returning an integer value which is the
+ *    number of milliseconds for which the site will wait between printing
+ *    successive tokens.
+ */
+const slowFill = async (entity, text, wait_function) => {
+  /*
+    Boolean value indicating if there is a cursor attached to the
+    end of the text field being written to.
+  */
```
[the remark this answers is not in the document]
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js:43]: Cool, thanks for the feedback 👍 I'll fix this now

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -13,21 +13,118 @@
...
+  let cursor_attached = false;
 
-  // Add it to the page.
-  const factContainer = document.getElementById('fact-container');
-  factContainer.innerText = fact;
+  /* 
+    wait_function must be non-null.
+    all null inputs are normalised to a known default.
+   */
+  if (wait_function == null) wait_function = function() {50};
```
[the remark this answers is not in the document]
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js:43]: i've fixed this now 👍 https://github.com/thomasogara/step-repository/pull/2/commits/a54a7fdfca992950008641755503d37de23c8788

File: portfolio/src/main/webapp/script.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js]: fixed now 👍 https://github.com/thomasogara/step-repository/pull/2/commits/a54a7fdfca992950008641755503d37de23c8788

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -13,21 +13,118 @@
...
+  const wait_function = () => (Math.floor(Math.random() * 50) + 50);
+  slowFill(output_element, biography_text, wait_function);
+}
 
-  // Pick a random greeting.
-  const fact = facts[Math.floor(Math.random() * facts.length)];
+/**
+ * Fill the text field of a given entity with a given string of text, character
+ * by character, at a programmable rate.
+ * @param entity The entity to write to the text field of.
```
[the remark this answers is not in the document]
Comment (thomasogara) [reply portfolio/src/main/webapp/script.js]: Cool, thanks for the feedback 👍 I added type info for all 3 params https://github.com/thomasogara/step-repository/pull/2/commits/2ac81f151f05403d0c828b1bd85f252d104b3676

**Pull request merged**